packages/react/src/combobox/root/AriaCombobox.tsx

```diff
@@ -857,6 +689,183 @@ export function AriaCombobox<Value = any, Mode extends SelectionMode = 'none'>(
+      store.state.onItemHighlighted(
+        nextActiveValue,
+        createGenericEventDetails('none', undefined, { index: storeActiveIndex }),
+      );
+    }
+  }, [activeIndex, autoHighlight, flatFilteredItems, inline, items, open, store, valuesRef]);
+
+  // When the available items change, ensure the selected value(s) remain valid.
+  // - Single: if current selection is removed, fall back to defaultSelectedValue if it exists in the list; else null.
+  // - Multiple: drop any removed selections.
+  useIsoLayoutEffect(() => {
+    if (!items || selectionMode === 'none') {
+      return;
+    }
+
+    const registry = flatItems;
+
+    if (multiple) {
+      const current = Array.isArray(selectedValue) ? selectedValue : EMPTY_ARRAY;
+      const next = current.filter((v) => itemIncludes(registry, v, store.state.isItemEqualToValue));
+      if (next.length !== current.length) {
+        setSelectedValueUnwrapped(next);
+      }
+      return;
+    }
+
+    const isStillPresent =
+      selectedValue == null
+        ? true
+        : itemIncludes(registry, selectedValue, store.state.isItemEqualToValue);
+    if (isStillPresent) {
+      return;
+    }
+
+    let fallback = null;
+    if (
+      defaultSelectedValue != null &&
+      itemIncludes(registry, defaultSelectedValue, store.state.isItemEqualToValue)
+    ) {
+      fallback = defaultSelectedValue;
+    }
+    setSelectedValueUnwrapped(fallback);
+
+    // Keep the input text in sync when the input is rendered outside the popup.
+    if (!store.state.inputInsidePopup) {
+      const stringVal = stringifyAsLabel(fallback, itemToStringLabel);
+      if (inputRef.current && inputRef.current.value !== stringVal) {
+        setInputValueUnwrapped(stringVal);
+      }
+    }
+  }, [
+    items,
+    flatItems,
+    multiple,
+    selectionMode,
+    selectedValue,
+    defaultSelectedValue,
+    setSelectedValueUnwrapped,
+    setInputValueUnwrapped,
+    itemToStringLabel,
+    store,
+  ]);
+
+  useValueChanged(queryRef, query, () => {
+    if (!open || query === '' || query === String(initialDefaultInputValue)) {
+      return;
+    }
+    setQueryChangedAfterOpen(true);
+  });
+
+  useValueChanged(selectedValueRef, selectedValue, (previousSelectedValue) => {
+    if (selectionMode === 'none') {
+      return;
+    }
+
+    clearErrors(name);
+    commitValidation?.(selectedValue, true);
+
+    if (validationMode === 'onChange') {
+      commitValidation?.(selectedValue);
+    }
+
+    updateValue(selectedValue);
+
+    if (selectionMode === 'single' && !hasInputValue && !inputInsidePopup) {
+      const nextInputValue = stringifyAsLabel(selectedValue, itemToStringLabel);
+      const previousInputValue = stringifyAsLabel(previousSelectedValue, itemToStringLabel);
+
+      if (nextInputValue !== previousInputValue && inputValueRef.current !== nextInputValue) {
+        setInputValue(nextInputValue, createChangeEventDetails('none'));
+      }
+    }
```

Comment:
Actual diff. I moved the effects down since this needs to access `setInputValue`, and all the similar types internal data are located next to each other intentionally (setters vs. effects, etc)